solr/core/src/java/org/apache/solr/util/circuitbreaker/GlobalCircuitBreakerManager.java

```diff
@@ -0,0 +1,156 @@
+import java.util.concurrent.ConcurrentHashMap;
+import org.apache.solr.client.solrj.SolrRequest;
+import org.apache.solr.common.annotation.JsonProperty;
+import org.apache.solr.common.cloud.ClusterPropertiesListener;
+import org.apache.solr.common.util.Utils;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.util.SolrJacksonAnnotationInspector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class GlobalCircuitBreakerManager implements ClusterPropertiesListener {
+  private static final Logger log = LoggerFactory.getLogger(MethodHandles.lookup().lookupClass());
+  private static final ObjectMapper mapper = SolrJacksonAnnotationInspector.createObjectMapper();
+  private final CircuitBreakerRegistry cbRegistry = new CircuitBreakerRegistry();
+  private final GlobalCircuitBreakerFactory factory;
+  private final CoreContainer coreContainer;
+  private int currentConfigHash;
+
+  public CircuitBreakerRegistry getCircuitBreakerRegistry() {
+    return cbRegistry;
+  }
+
+  public GlobalCircuitBreakerManager(CoreContainer coreContainer) {
+    super();
+    this.factory = new GlobalCircuitBreakerFactory(coreContainer);
+    this.coreContainer = coreContainer;
+  }
+
+  private static class GlobalCircuitBreakerConfig {
+    static final String CIRCUIT_BREAKER_CLUSTER_PROPS_KEY = "circuit-breakers";
+    @JsonProperty Map<String, CircuitBreakerConfig> configs = new ConcurrentHashMap<>();
+
+    @JsonProperty
+    Map<String, Map<String, CircuitBreakerConfig>> hostOverrides = new ConcurrentHashMap<>();
+
+    static class CircuitBreakerConfig {
+      @JsonProperty Boolean enabled = false;
+      @JsonProperty Boolean debugMode = false;
+      @JsonProperty Double updateThreshold = Double.MAX_VALUE;
+      @JsonProperty Double queryThreshold = Double.MAX_VALUE;
+    }
+  }
+
+  // for registering global circuit breakers set in clusterprops
+  @Override
+  public boolean onChange(Map<String, Object> properties) {
+    try {
+      GlobalCircuitBreakerConfig nextConfig = processConfigChange(properties);
+      if (nextConfig != null) {
+        registerCircuitBreakers(nextConfig);
+      }
+    } catch (Exception e) {
+      if (log.isWarnEnabled()) {
+        // don't break when things are misconfigured
+        log.warn("error parsing global circuit breaker configuration {}", e);
+      }
+    }
+    return false;
+  }
+
+  private GlobalCircuitBreakerConfig processConfigChange(Map<String, Object> properties)
+      throws IOException {
+    Object cbConfig = properties.get(GlobalCircuitBreakerConfig.CIRCUIT_BREAKER_CLUSTER_PROPS_KEY);
+    GlobalCircuitBreakerConfig globalCBConfig = null;
+    if (cbConfig != null) {
+      byte[] configInput =
+          Utils.toJSON(
+              properties.get(GlobalCircuitBreakerConfig.CIRCUIT_BREAKER_CLUSTER_PROPS_KEY));
+
+      int nextHashCode = Arrays.hashCode(configInput);
+      if (this.currentConfigHash == nextHashCode) {
+        return null;
+      }
```

Comment:
Ah okay, I'll re-evaluate the approach here. I did originally try overriding equals and hashCode, but the code got kind of cumbersome. Maybe now that it's all ironed out otherwise I can try that again, and default to the cryptographic hash otherwise.